auth_saml/controllers/main.py

```diff
@@ -0,0 +1,210 @@
+    def wrapper(self, req, **kw):
+        if not kw:
+            return """<html><head><script>
+                var l = window.location;
+                var q = l.hash.substring(1);
+                var r = '/' + l.search;
+                if(q.length !== 0) {
+                    var s = l.search ? (l.search === '?' ? '' : '&') : '?';
+                    r = l.pathname + l.search + s + q;
+                }
+                window.location = r;
+            </script></head><body></body></html>"""
+        return func(self, req, **kw)
+    return wrapper
+
+
+# ----------------------------------------------------------
+# Controller
+# ----------------------------------------------------------
+
+
+class SAMLLogin(Home):
+
+    def list_providers(self):
+        providers = request.env['auth.saml.provider'].sudo().search_read([])
+        for provider in providers:
+            # Compatibility with auth_oauth/controllers/main.py in order to
+            # avoid KeyError rendering template_auth_oauth_providers
+            provider.setdefault('auth_link', "")
+        return providers
+
+    @http.route()
+    def web_login(self, *args, **kw):
+        ensure_db()
+        if (
+            request.httprequest.method == 'GET' and
+            request.session.uid and
+            request.params.get('redirect')
+        ):
+
+            # Redirect if already logged in and redirect param is present
+            return http.redirect_with_hash(request.params.get('redirect'))
+
+        providers = self.list_providers()
+
+        response = super(SAMLLogin, self).web_login(*args, **kw)
+        if response.is_qweb:
+            error = request.params.get('saml_error')
+            if error == 'no-signup':
+                error = _("Sign up is not allowed on this database.")
+            elif error == 'access-denied':
+                error = _("Access Denied")
+            elif error == 'expired':
+                error = _(
+                    "You do not have access to this database or your "
+                    "invitation has expired. Please ask for an invitation "
+                    "and be sure to follow the link in your invitation email."
+                )
+            else:
+                error = None
+
+            response.qcontext['providers'] = providers
+
+            if error:
+                response.qcontext['error'] = error
+
+        return response
+
+
+class AuthSAMLController(http.Controller):
+
+    def get_state(self, provider_id):
+        """Compute a state to be sent to the IDP so it can forward it back to
+        us.
+
+        :rtype: Dictionary.
+        """
+
+        redirect = request.params.get('redirect') or 'web'
+        if not redirect.startswith(('//', 'http://', 'https://')):
+            redirect = '%s%s' % (
+                request.httprequest.url_root,
+                redirect[1:] if redirect[0] == '/' else redirect
+            )
+
+        state = {
+            "d": request.session.db,
```

Comment:
Can be good use other key names :/

Reply:
I think this is required by the SAML standard. It's being used as raw JSON somewhere.